Proposal: Vector (#2587)
swiftlang/swift-evolution: proposals/nnnn-vector.md

# Vector, a fixed-size array

* Proposal: [SE-NNNN](NNNN-vector.md)
* Authors: [Alejandro Alonso](https://github.com/Azoy)
* Review Manager: TBD
* Status: **Awaiting implementation**

> * Roadmap: [Approaches for fixed-size arrays](https://forums.swift.org/t/keypath-equality-issue-where-root-is-a-generic-type/14865/12)
> * Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN)

Reviewer: ```suggestion
* Roadmap: [Approaches for fixed-size arrays](https://forums.swift.org/t/approaches-for-fixed-size-arrays/58894)
* Implementation: [swiftlang/swift#76438](https://github.com/swiftlang/swift/pull/76438)
```

Author: I don't know how I managed to do this, thanks!

fixed-size contiguously inline allocated array. We're defining "inline" as using
the most natural allocation pattern depending on the context of where this is
used. It will be stack allocated most of the time, but as a class property
member it will be inline allocated on the heap with the rest of the properties.
`Vector` will never introduce an implicit heap allocation just for its storage
alone.

```swift
func complexAlgorithm() -> Int {
  // This is a stack allocation, no 'malloc's or reference counting here!
  let elements: Vector<4, Int> = [1, 2, 3, 4]

  // OK
  for element in elements {
    compute(element)
  }

  var i = 0

  while i != elements.indices {
    let element = elements[i] // OK

    compute(element)
  }
}
```

This type also serves as a safer replacement to `withUnsafeTemporaryAllocation`
when the amount of things to reserve is known statically:

```swift
// Previously
withUnsafeTemporaryAllocation(of: Int.self, capacity: 16) {
  $0.initialize(repeating: 0)

  for i in $0.indices {
    $0[i] = compute()
  }
}

// Now
var ints = Vector<16, Int>(repeating: 0)

for i in ints.indices {
  ints[i] = compute()
}
```

## Detailed design

> `Vector` will be a simple noncopyable struct capable of storing other potentialy

Reviewer: ```suggestion
`Vector` will be a simple noncopyable struct capable of storing other potentially
```

noncopyable elements. It will be conditionally copyable only when its elements
are.

```swift
public struct Vector<let Count: Int, Element: ~Copyable>: ~Copyable {}

extension Vector: Copyable where Element: Copyable {}
extension Vector: BitwiseCopyable where Element: BitwiseCopyable {}
extension Vector: Sendable where Element: Sendable {}
```

### MemoryLayout

The memory layout of a `Vector` is defined by taking its `Element`'s stride and

> multiplying that by its `Count` for its size and stride. It's alignment is equal

Reviewer 2: ```suggestion
multiplying that by its `Count` for its size and stride. Its alignment is equal
```

  /// the index of the current element being initialized. The closure is allowed
  /// to throw an error at any point during initialization at which point the
  /// vector will stop initialization, deinitialize every currently initialized
  /// element, and throw the given error back out to the caller.
  ///
  /// - Parameter body: A closure that returns an owned `Element` to emplace at
  ///                   the passed in index.
  public init<E: Error>(_ body: (Int) throws(E) -> Element) throws(E)
}

extension Vector where Element: Copyable {
  /// Initializes every element in this vector to a copy of the given value.
  ///
  /// - Parameter value: The instance to initialize this vector with.
  public init(repeating: Element)
}
```

### Deinitialization and consumption

Once a vector is no longer used, the compiler will implicitly destroy its value.
This means that it will do an element by element deinitialization, releasing any
class references or calling any `deinit`s on noncopyable types. Doing
deinitialization this way will lose all access to every element. If one is done
with the vector value but would like retain one or more of its elements, then a

> call to the `destroy` method will grant access to the owned element value for

Reviewer 3: This says "destroy", but the method is called "consume" in the snippet.

  /// remaining elements and throw the given error back out to the caller.
  ///
  /// - Parameter body: A closure given the current deinitialized element's
  ///                   index and the owned element value.
  /// - Returns: The return value, if any, of the `body` closure parameter.
  public consuming func consume<Result: ~Copyable, E: Error>(
    _ body: (Int, consuming Element) throws(E) -> Result
  ) throws(E) -> Result
}
```

### Generalized `Collection` APIs

```swift
extension Vector where Element: ~Copyable {
  public typealias Element = Element
  public typealias Index = Int
  public typealias Indices = Range<Int>

  public var count: Int { Count }
  public var indices: Range<Int> { 0 ..< Count }
  public var isEmpty: Bool { Count == 0 }
  public var startIndex: Int { 0 }
  public var endIndex: Int { Count }

  public borrowing func index(after i: Int) -> Int
  public borrowing func index(before i: Int) -> Int

  public subscript(_ index: Int) -> Element
}
```

### `Collection` conformances for copyable elements

Here, we define our basic collection conformances to `RandomAccessCollection`,

> `MutableCollection`, and `BidirectionalCollection` but onlt when the element is

Reviewer 3: "onlt"

  let offset: UInt32
  let align: UInt32
  ...
}
```

If the C struct already contained a field named `sectnameVector`, then we
wouldn't generate a `Vector` field for `sectname`. However, one should be able
to mitigate this by annotating the C field directly with `swift_vector_name`:

```c
#define SWIFT_VECTOR_NAME(X) __attribute__((__swift_vector_name__(#X)))

struct section_64 {
  SWIFT_VECTOR_NAME(sectname_vector)
  char sectname[16];

  SWIFT_VECTOR_NAME(segname_vector)
  char segname[16];

  uint64_t addr;
  uint64_t size;
  uint32_t offset;
  uint32_t align;
  ...
};
```

Note, that while the imported C struct looks like it has more members, its
memory layout will be unaffected.

In a future Swift language mode, we plan to disable this behavior and drop
importing C array fields as tuples and only as `Vector` with the same name from
the C struct. This approach lets folks who are ready to migrate their codebase
using a C interface to use `Vector` entirely, while still allowing dependencies
or dependents to use the old behavior with the same C interface.

## Source compatibility

`Vector` is a brand new type in the standard library, so source should still be
compatible.

> Given the name of this type however, we forsee this clashing with existing user

Reviewer: ```suggestion
Given the name of this type however, we foresee this clashing with existing user
```

defined types named `Vector`. This isn't a particular issue though because the
standard library has special shadowing rules which prefer user defined types by
default. Which means in user code with a custom `Vector` type, that type will

> always be preferred over the standard library's `Swift.Vector`.

Reviewer 2: I'm assuming it does, but can you clarify if "always" means that `MyLib.Vector<T>` will be preferred over `Swift.Vector<N, T>` even if I write `Vector<3, Int>`?

Author: This is a good find and it seems the behavior is yes, it will \_always\_ be preferred. I'll make a note of it in the proposal.

compatible.

## Implications on adoption

This is a brand new type which means there will be deployment version
requirement to be able to use this type, especially considering it is using new
runtime features from integer generics.

## Future directions

### `Equatable`, `Hashable`, `CustomStringConvertible`, and other protocols.

There are a wide class of protocols that this type has the ability to conform to,
but the issue is that it can only conform to them when the element conforms to
them (this is untrue for `CustomStringConvertible`, but it still requires
copyability). We could introduce these conformances but have them be conditional
right now and generalize it later when we generalize these protocols, but if we
were to ship say Swift X.Y with:

```swift
@available(SwiftStdlib X.Y)
extension Vector: Equatable where Element: Equatable // & Element: Copyable
```

and later down the road in Swift X.(Y + 1):

```swift
@available(SwiftStdlib X.Y)
extension Vector: Equatable where Element: ~Copyable & Equatable
```

Suddenly, this availability isn't quite right because the conformance that
shipped in Swift X.Y doesn't support noncopyable elements. To prevent the

> headache of this an any potential new availability feature, we're holding off on

Reviewer: ```suggestion
headache of this and any potential new availability feature, we're holding off on
```

```suggestion
headache of this on any potential new availability feature, we're holding off on
```

these conformances until they are fully generalized.

### Noncopyable `Sequence` and `Collection` protocols

Similarly, while we aren't conforming to a lot of protocols that we envision
we'll generalize one day, we are conforming to `Sequence`, `Collection`,

> `RandomAccessCollection`, and `MutableCollection`. We don't forsee that we'll

Reviewer: ```suggestion
`RandomAccessCollection`, and `MutableCollection`. We don't foresee that we'll
```

these types:

```swift
public protocol ExpressibleByVectorLiteral: ~Copyable {
  associatedtype Element: ~Copyable
  associatedvalue Count: Int

  init(vectorLiteral: consuming Vector<Count, Element>)
}

extension Vector: ExpressibleByVectorLiteral {
  init(vectorLiteral: consuming Vector<Count, Element>) { ... }
}

extension Pair: ExpressibleByVectorLiteral {
  init(vectorLiteral: consuming Vector<2, Element>) { ... }
}
```

But even with this design it's unsuitable for `Array` itself because it doesn't
want a static count for the literal, it still wants it to be generic.

It would be nice to define something like this either on top of `Vector`,
parameter packs, or something else that would let us define statically the
number of elements we need for literal initialization or be dynamic if we opt to.

### `InlineArray` and `SmallArray`

In the same vein as this type, it may make sense to introduce some `InlineArray`
type which would support appending and removing elements given a fixed-capacity.

```swift
var numbers: InlineArray<4, Int> = [1, 2]
print(numbers.capacity) // 4
print(numbers.count) // 2
numbers.append(3)
print(numbers.count) // 3

> numers.append(4)

Reviewer: ```suggestion
numbers.append(4)
```

capacity and would fall back to a dynamic heap allocation.

### Syntax sugar

We feel that this type will become as fundamental as `Array` and `Dictionary`
both of which have syntactic sugar for declaring a type of them, `[T]` for
`Array` and `[K: V]` for `Dictionary`. It may make sense to define something
similar for `Vector`, however we leave that as a future direction as the
spelling for such syntax is not critical to landing this type.

It should be fairly trivial to propose such a syntax in the future either via a
new proposal, or as an amendment to this one. Such a change should only require
a newer compiler that supports the syntax and nothing more.

Some syntax suggestions:

* `[N x T]` or `[T x N]`
* `[N * T]` or `[T * N]`
* `T[N]` (from C)
* `[T; N]` (from Rust)

Note that it may make more sense to have the length appear before the type. I
discuss this more in depth in [Reorder the generic arguments](#reorder-the-generic-arguments-vectort-n-instead-of-vectorn-t).

## Alternatives considered

### A name other than `Vector`

For obvious reasons, we cannot name this type `Swift.Array` to match the
"term of art" that other languages like C, C++, and Rust are using for this

> exact type. However, while this name is the defacto for other languages, it

Reviewer: ```suggestion
exact type. However, while this name is the de facto for other languages, it
```

and [Rust's arrays `[T; N]`](https://doc.rust-lang.org/std/primitive.array.html).

## Motivation

Arrays in Swift have served as the go to choice when needing to put items in an
ordered list. They are a great data structure ranging from a variety of
different use cases from teaching new developers all the way up to sophisticated
implementation details of something like a cache.

However, using `Array` all the time doesn't really make sense in some scenarios.
It's important to understand that `Array` is a heap allocated growable data
structure which can be expensive and unnecessary in some situations. The next
best thing is to force a known quantity of elements onto the stack, probably by
using tuples.

```swift
func complexAlgorithm() {
  let elements = (first, second, third, fourth)
}
```

Unfortunately, using tuples in this way is very limited. They don't allow for
dynamic indexing or iteration:

```swift
func complexAlgorithm() {
  let elements = (first, second, third, fourth)

  // error: for-in loop requires '(Int, Int, Int, Int)' to conform to 'Sequence'
  for element in elements {
    compute(element)

>   }

Reviewer 4: We should update this and the proposed solution snippet below to account for the fact that `Vector` cannot be iterated directly anymore.

Author: done!

  var i = 0

>   while i != elements.indices {

Reviewer 4: This should be `elements.indices.count`, and we should have an `i += 1` somewhere, right?

Reviewer 4: (Same below)

Author: Fixed these examples to just use a for loop over the indices.

```swift
@available(SwiftStdlib X.Y)
extension Vector: Equatable where Element: ~Copyable & Equatable
```

Suddenly, this availability isn't quite right because the conformance that
shipped in Swift X.Y doesn't support noncopyable elements. To prevent the
headache of this and any potential new availability feature, we're holding off on
these conformances until they are fully generalized.

### `Sequence` and `Collection`

Similarly, we aren't conforming to `Sequence` or `Collection` either.
While we could conform to these protocols when the element is copyable, `Vector`
is unlike `Array` in that there are no copy-on-write semantics; it is eagerly
copied. Conforming to these protocols would potentially open doors to lots of
implicit copies of the underlying vector instance which could be problematic
given the prevalence of generic collection algorithms and slicing behavior. To
avoid this potential performance pitfall, we're explicitly not opting into
conforming this type to `Sequence` or `Collection`.

We do plan to propose new protocols that look like `Sequence` and `Collection`
that avoid implicit copying making them suitable for types like `Vector` and

> containers of noncopyble elements.

Reviewer: ```suggestion
containers of noncopyable elements.
```
